File: results/dnd-rest-benchmark/lifecycle-runs/20260710T233736Z_pi_kimi-k2p7-code_typescript-nextjs/app/lib/compendium.ts
import { getDb } from "./db.js";

export interface Monster {
  slug: string;
  name: string;
  cr: string;
  armor_class: number;
  hit_points: number;
  tags: string[];
}

export interface Item {
  slug: string;
  name: string;
  type: string;
  rarity: string;
  cost_gp: number;
}

export function createMonster(
  input: Record<string, unknown>
): Omit<Monster, "tags"> {
  const slug = validateSlug(input.slug);
  const name = validateName(input.name);
  const cr = validateString(input.cr, "cr");
  const armor_class = validateInteger(input.armor_class, "armor_class");
  const hit_points = validateInteger(input.hit_points, "hit_points");
  const tags = validateStringArray(input.tags, "tags");

  const db = getDb();
  const existing = db
    .prepare("SELECT slug FROM monsters WHERE slug = ?")
    .get(slug);
  if (existing) {
    throw new Error("monster slug already exists");
  }

  db.prepare(
    `
    INSERT INTO monsters (slug, name, cr, armor_class, hit_points, tags_json)
    VALUES (?, ?, ?, ?, ?, ?)
    `
  ).run(slug, name, cr, armor_class, hit_points, JSON.stringify(tags));

  return { slug, name, cr, armor_class, hit_points };
}

export function getMonster(slug: string): Monster | undefined {
  const db = getDb();
  const row = db
    .prepare(
      `
      SELECT slug, name, cr, armor_class, hit_points, tags_json
      FROM monsters WHERE slug = ?
      `
    )
    .get(slug) as Record<string, unknown> | undefined;
  if (!row) return undefined;
  return {
    slug: String(row.slug),
    name: String(row.name),
    cr: String(row.cr),
    armor_class: Number(row.armor_class),
    hit_points: Number(row.hit_points),
    tags: JSON.parse(String(row.tags_json)),
  };
}

export function createItem(input: Record<string, unknown>): Item {
  const slug = validateSlug(input.slug);
  const name = validateName(input.name);
  const type = validateString(input.type, "type");
  const rarity = validateString(input.rarity, "rarity");
  const cost_gp = validateInteger(input.cost_gp, "cost_gp");

  const db = getDb();
  const existing = db.prepare("SELECT slug FROM items WHERE slug = ?").get(slug);
  if (existing) {
    throw new Error("item slug already exists");
  }

  db.prepare(
    `
    INSERT INTO items (slug, name, type, rarity, cost_gp)
    VALUES (?, ?, ?, ?, ?)
    `
  ).run(slug, name, type, rarity, cost_gp);

  return { slug, name, type, rarity, cost_gp };
}

export function getItem(slug: string): Item | undefined {
  const db = getDb();
  const row = db
    .prepare(
      `
      SELECT slug, name, type, rarity, cost_gp
      FROM items WHERE slug = ?
      `
    )
    .get(slug) as Record<string, unknown> | undefined;
  if (!row) return undefined;
  return {
    slug: String(row.slug),
    name: String(row.name),
    type: String(row.type),
    rarity: String(row.rarity),
    cost_gp: Number(row.cost_gp),
  };
}

function validateSlug(value: unknown): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new Error("slug must be a non-empty string");
  }
  return value;
}

function validateName(value: unknown): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new Error("name must be a non-empty string");
  }
  return value;
}

function validateString(value: unknown, field: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new Error(`${field} must be a non-empty string`);
  }
  return value;
}

function validateInteger(value: unknown, field: string): number {
  if (typeof value !== "number" || !Number.isInteger(value)) {
    throw new Error(`${field} must be an integer`);
  }
  return value;
}

function validateStringArray(value: unknown, field: string): string[] {
  if (!Array.isArray(value)) {
    throw new Error(`${field} must be an array`);
  }
  for (const item of value) {
    if (typeof item !== "string") {
      throw new Error(`${field} must be an array of strings`);
    }
  }
  return value;
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075409Z_pi_kimi-k2p7-code_ruby-stdlib/lib/compendium.rb
# frozen_string_literal: true

require 'json'
require_relative 'persistence'

# Compendium: monsters and items.
module Compendium
  def self.create_monster(payload)
    data = validate_monster(payload)
    return [:invalid] unless data

    Persistence.db do |d|
      existing = d.get_first_value('SELECT 1 FROM compendium_monsters WHERE slug = ?', data[:slug])
      next [:conflict] if existing

      d.execute(
        'INSERT INTO compendium_monsters (slug, name, cr, armor_class, hit_points, tags_json) VALUES (?, ?, ?, ?, ?, ?)',
        [data[:slug], data[:name], data[:cr], data[:armor_class], data[:hit_points], JSON.generate(data[:tags])]
      )

      [:ok, {
        'slug' => data[:slug],
        'name' => data[:name],
        'cr' => data[:cr],
        'armor_class' => data[:armor_class],
        'hit_points' => data[:hit_points]
      }]
    end
  end

  def self.read_monster(slug)
    return [:invalid] unless slug.is_a?(String) && !slug.empty?

    Persistence.db do |d|
      row = d.get_first_row(
        'SELECT name, cr, armor_class, hit_points, tags_json FROM compendium_monsters WHERE slug = ?',
        slug
      )
      next [:not_found] unless row

      name, cr, armor_class, hit_points, tags_json = row
      tags = JSON.parse(tags_json) rescue []

      [:ok, {
        'slug' => slug,
        'name' => name,
        'cr' => cr,
        'armor_class' => armor_class,
        'hit_points' => hit_points,
        'tags' => tags
      }]
    end
  end

  def self.create_item(payload)
    data = validate_item(payload)
    return [:invalid] unless data

    Persistence.db do |d|
      existing = d.get_first_value('SELECT 1 FROM compendium_items WHERE slug = ?', data[:slug])
      next [:conflict] if existing

      d.execute(
        'INSERT INTO compendium_items (slug, name, type, rarity, cost_gp) VALUES (?, ?, ?, ?, ?)',
        [data[:slug], data[:name], data[:type], data[:rarity], data[:cost_gp]]
      )

      [:ok, {
        'slug' => data[:slug],
        'name' => data[:name],
        'type' => data[:type],
        'rarity' => data[:rarity],
        'cost_gp' => data[:cost_gp]
      }]
    end
  end

  def self.read_item(slug)
    return [:invalid] unless slug.is_a?(String) && !slug.empty?

    Persistence.db do |d|
      row = d.get_first_row(
        'SELECT name, type, rarity, cost_gp FROM compendium_items WHERE slug = ?',
        slug
      )
      next [:not_found] unless row

      name, type, rarity, cost_gp = row

      [:ok, {
        'slug' => slug,
        'name' => name,
        'type' => type,
        'rarity' => rarity,
        'cost_gp' => cost_gp
      }]
    end
  end

  def self.validate_monster(payload)
    return nil unless payload.is_a?(Hash)

    slug = payload['slug']
    name = payload['name']
    cr = payload['cr']
    armor_class = payload['armor_class']
    hit_points = payload['hit_points']
    tags = payload['tags']

    return nil unless slug.is_a?(String) && !slug.empty? &&
                      name.is_a?(String) && !name.empty? &&
                      cr.is_a?(String) && !cr.empty? &&
                      armor_class.is_a?(Integer) &&
                      hit_points.is_a?(Integer) &&
                      tags.is_a?(Array) &&
                      tags.all? { |t| t.is_a?(String) }

    {
      slug: slug,
      name: name,
      cr: cr,
      armor_class: armor_class,
      hit_points: hit_points,
      tags: tags
    }
  end
  private_class_method :validate_monster

  def self.validate_item(payload)
    return nil unless payload.is_a?(Hash)

    slug = payload['slug']
    name = payload['name']
    type = payload['type']
    rarity = payload['rarity']
    cost_gp = payload['cost_gp']

    return nil unless slug.is_a?(String) && !slug.empty? &&
                      name.is_a?(String) && !name.empty? &&
                      type.is_a?(String) && !type.empty? &&
                      rarity.is_a?(String) && !rarity.empty? &&
                      cost_gp.is_a?(Integer)

    {
      slug: slug,
      name: name,
      type: type,
      rarity: rarity,
      cost_gp: cost_gp
    }
  end
  private_class_method :validate_item
end
